controller.py

```diff
@@ -81,46 +83,157 @@
+        """Look up the 'unaffiliated' collection, which is used to register
+        identifiers referenced by unauthenticated clients in one-off tests.
+        """
+        if self._default_collection_id is None:
+            default_collection, ignore = IdentifierResolutionCoverageProvider.unaffiliated_collection(self._db)
+            self._default_collection_id = default_collection.id
+        return get_one(self._db, Collection, id=self._default_collection_id)
+
+    def load_collection(self, metadata_identifier, authentication_required=True):
+        """Load or create a Collection from details provided in an incoming
+        request.
+
+        :param metadata_identifier: The metadata identifier for a Collection.
+
+        :param authentication_required: If this is True, then
+               unauthenticated or improperly authenticated requests
+               will result in a ProblemDetail. If this is False, then
+               unauthenticated requests are allowed, but they will always
+               be directed to the 'unaffiliated' collection.
+
+        :return: A Collection for the given metadata identifier, if
+                 one was found or created. A ProblemDetail, if no
+                 IntegrationClient could be authorized, or if the
+                 metadata identifier was invalid. None, if anonymous
+                 access is allowed and no authentication was provided.
+        """
+        # Only an authenticated IntegrationClient has any reason to
+        # look at a specific collection.
+        client = MetadataWrangler.authenticated_client_from_request(
+            self._db, authentication_required
+        )
+        if isinstance(client, ProblemDetail):
+            return client
+
+        if not client:
+            if authentication_required:
+                # If you authenticate, you must mention a specific
+                # Collection.
+                if not metadata_identifier:
+                    return INVALID_INPUT.detailed(_("No metadata identifier provided."))
+            else:
+                # If you don't authenticate, you can't load or create a
+                # Collection. You always get the default collection.
+                return self.default_collection
+
+        # For collections with the ExternalIntegration.OPDS_IMPORT
+        # protocol, a DataSource as well as a metadata identifier may
+        # be sent.
+        data_source_name = flask.request.args.get('data_source')
+
+        # Load or create a Collection based on the metadata identifier
+        # and data source name.
+        try:
+            collection, ignore = Collection.from_metadata_identifier(
+                self._db, metadata_identifier, data_source=data_source_name
+            )
+        except ValueError as e:
+            # This is caused by a problem decoding the metadata
+            # identifier.
+            return INVALID_INPUT.detailed(unicode(e))
+        return collection
```

Comment:
Previously, if you passed in a bad metadata identifier, you'd get an exception that didn't explain what went wrong.